### ISSUE TITLE
line 156 in RootElement.js "var subscribe" function def not hoisted

### ISSUE THREAD

**Issue by jjunkin** [redfin/react-server#763]
In file react-server/packages/react-server/core/components/RootElement.js

on line 156: var subscribe = callback => {updater = callback};

"var subscribe" is declared and the variable is hoisted. Yet as far as I can tell the function definition is not hoisted. Therefore, any usage above the definition the variable "subscribe" will be undefined, even though it is declared.

The only place I see "subscribe" is used is line 151 which is above this line, therefore I believe the value it passes will always be "undefined".

It appears that "this.props.subscribe" is evaluated on line 14 and if it has a value it is used. As far as I can tell this is the same value and it is always undefined. Therefore, the "this.props.subscribe " on line 15 is never executed.

If I am wrong about this please let me know. It seems like a fairly critical issue if I am correct.

A possible solution would be to move the function expression definition of 
156: var subscribe = callback => {updater = callback};
above the definition of 
137: var unsubscribe = listen(childProps => {

**Comment by aickin** [comment]
I think that you're right that this isn't ideal, but practically it shouldn't cause a bug because the `subscribe` reference on line 151 is in a callback, so it happens after `subscribe` has been assigned. But you're right, the code would be cleaner with the definition above, and it'd be a great first PR to help the project. Would you be willing to submit it?

**Comment by jjunkin** [comment]
When I stepped through the code with the chrome debugger, subscribe was always undefined - including within the callback function. I realize this could be different on various browsers (using chrome on osx).

From my understanding the "listen={emitter}" functionality (documented in Understanding Rendering) depends on this subscribe function being defined, therefore does not work as documented. Using the "listen=" prop I am able to get data to load in the initial render but I am unable to trigger a data reload after that.

Do you have any examples of the "listen" prop working  after the initial data load? I was unable to find any in the codebase. I noticed the test PropsFromEmitterRootElement.js but this only tests the initial data load using a single setTimeout. Examples of this functionality seem critical for the project and would be helpful.

I would be happy to try and submit a change if I get the time next week. Thanks for the response!

**Comment by gigabo** [comment]
Oh, it actually looks like #769 addresses this. 👀

**Comment by aickin** [comment]
@jjunkin: Thanks for the info and for tracing through the code! I stand corrected.

@gigabo: how would you feel about activating the `no-use-before-define` ESLint rule? That rule would have caught this bug. I think the main downside would be that you would have to define functions in order in a file, even when defining them out of order will never practically cause a bug. For example, this code would error:

```
// this would cause a linting error.
export default x => addOne(x);
const addOne = y => y + 1;
```

But, as I say, the upside is that this bug would have been caught.

If you think this is a good idea, I'd be happy to implement the change and update any code that needs to be updated to lint correctly. Thoughts?

**Comment by gigabo** [comment]
> Do you have any examples of the "listen" prop working after the initial data load?

Yeah, it works well for us internally at Redfin, but this does look buggy.  Specifically this looks like it will fail if `listen` calls back synchronously.

@aickin I fear you'll find a lot of `no-use-before-define` violations. 😬 

But if you're willing to take it on and it can help us avoid this sort of subtle bug then I'm all for it!

**Comment by gigabo** [comment]
Okay, I think #769 took care of this, and #770 should prevent this whole class of issues from happening again.  Available in [`v0.5.0`](https://github.com/redfin/react-server/releases/tag/v0.5.0).  Thanks!
